yt/yt/library/containers/cri/image_cache.cpp

```diff
@@ -0,0 +1,482 @@
+                    YT_LOG_DEBUG("Unmanaged docker image found in cache (Image: %v, ImageId: %v)",
+                        image,
+                        imageId);
+                    return MakeFuture(entry);
+                }
+                auto cookie = BeginInsert(image.Image);
+                auto imageFuture = cookie.GetValue();
+                if (cookie.IsActive()) {
+                    DoInsertImage(cookie, protoImage);
+                }
+                return imageFuture;
+            }));
+    }
+
+    TFuture<TCriImageCacheEntryPtr> PullImage(const TCriImageDescriptor& image, TCriAuthConfigPtr authConfig, EImagePullPolicy pullPolicy) override
+    {
+        if (!IsManagedImage(image)) {
+            // Bypass cache for unmanaged images.
+            return Executor_->GetImageStatus(image)
+                .Apply(BIND([=] (const TCriImageApi::TRspImageStatusPtr& imageStatus) {
+                    if (imageStatus->has_image()) {
+                        auto protoImage = imageStatus->image();
+                        TCriImageDescriptor imageId{.Image = protoImage.id()};
+                        YT_LOG_DEBUG("Unmanaged docker image found in cache (Image: %v, ImageId: %v)",
+                            image,
+                            imageId);
+                        return New<TCriImageCacheEntry>(imageId, image, protoImage.size());
+                    } else {
+                        THROW_ERROR_EXCEPTION("Unmanaged docker image not found in cache (Image: %v)",
+                            image);
+                    }
+                }));
+        }
+
+        switch (pullPolicy) {
+            case EImagePullPolicy::Never:
+                return LookupImage(image);
+            case EImagePullPolicy::Missing:
+                break;
+            case EImagePullPolicy::Default:
+                if (Config_->AlwaysPullLatest && image.Image.EndsWith(":latest")) {
+                    pullPolicy = EImagePullPolicy::Always;
+                }
+
+                if (Config_->PullPeriod) {
+                    auto imageEntry = Find(image.Image);
+                    if (imageEntry && imageEntry->GetLastPullTime() + Config_->PullPeriod < TInstant::Now()) {
+                        pullPolicy = EImagePullPolicy::Always;
+                    }
+                }
+
+                if (pullPolicy != EImagePullPolicy::Always) {
+                    break;
+                }
+                [[fallthrough]];
+            case EImagePullPolicy::Always:
+                // Should not remove incomplete entries.
+                TryRemove(image.Image, /*forbidResurrection*/ true);
+                break;
+        }
+
+        auto cookie = BeginInsert(image.Image, Config_->ImageSizeEstimation);
+
+        if (!cookie.IsActive()) {
+            auto imageFuture = cookie.GetValue();
+
+            if (auto entryOrError = imageFuture.TryGet(); entryOrError && entryOrError->IsOK()) {
```

Comment:
auto& ?

Reply:
where? I don't see here references

Reply:
auto& entryOrError
No need to copy error

Reply:
TryGet returns std::optional not reference
https://github.com/ytsaurus/ytsaurus/blob/0def99575af0e0fc4fdc04fbb46dbcc77085d31a/yt/yt/core/actions/future-inl.h#L271